### PR TITLE
Add Support for ONVIF cam as part of Ardupilot in SITL/Linux

### PR DESCRIPTION
**Pull request by bugobliterator**
Work on establishing all that is required for this to work on top of linux native networking, which will be later expanded to incorporate ChibiOS implementation again using LWIP.

Building and Running Test:
`cd modules/gsoap`
`autoreconf -f -i`
`./configure`
`make`
`make install #add sudo if needed` 
`./waf configure`
`./waf --target examples/onvif_test`
`./build/sitl/examples/onvif_test -M tracker`
Setting username pass and hotname can be done in `libraries/AP_ONVIF/AP_ONVIF.cpp`

After this you should see your camera move for one extreme to other

### CODE REVIEW

**Comment by rmackay9** [comment]
If this helps us getting better control of more cameras I'm excited!

**Comment by bugobliterator** [comment]
> what is the endgoal of this? to control an ONVIF camera ? How? via a mavlink bridge?

If all goes to plan, To control an ONVIF camera via AP_Camera and AP_Gimbal.

**Comment by bugobliterator** [comment]
ONVIF camera panning and tilting with antennatracker at the cost of 2MB extra, almost a 100% increase in size, the 18KB increase from my last PR doesn't look too big now does it @amilcarlucas . 😆

**Review by tridge — CHANGES_REQUESTED** [review]
I'm concerned about std string and vector handling on stm32, with memory allocation failure handling
it is worth looking at soaplua

File: libraries/AP_ONVIF/AP_ONVIF.cpp [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+
+// Generate Random Nonce value
+void AP_ONVIF::rand_nonce(char *nonce, size_t noncelen)
+{
+    size_t i;
+    uint32_t r = (uint32_t)(hal.util->get_hw_rtc()/1000000ULL);
+    (void)memcpy((void *)nonce, (const void *)&r, 4);
+    for (i = 4; i < noncelen; i += 4)
+    {
+        r = rand();
```
[the remark this answers is not in the document]
Comment (bugobliterator) [reply libraries/AP_ONVIF/AP_ONVIF.cpp:319]: this PR is introduced for the same: https://github.com/ArduPilot/ardupilot/pull/17855 , will rebase on that PR.

**Comment by bugobliterator** [comment]
Notes after doing investigations on points discussed on last call :

As we discussed on the call how memory handling is done by this library is of concern. Major issue was use of C++ STL. It turns out there is a very simple way to remove dependency on the std::string and vector, and replace them with simple dynamic arrays.

Next issue was how memory is being allocated by this library, after looking into it further it turns out the library does keep track of all memory it allocates and one can simply free all the memory at the end/failure in a transaction context with single call, much like talloc as I understood from @tridge 's explanation. Additionally the library has instrumentation to detect memory leakage, which I believe can be enabled in CI tests in the future.
There is also mechanism to override default memory allocation calls like malloc, which we can use to control memory allocation limits.

I do agree this is still not good enough to run in the same context as flight controller, but is definitely good enough to run in embedded system and is already doing so. It can be used in AP_Periph and other non-critical system/context.

Incase someone wants to read more on the gsoap library: https://www.genivia.com/doc/databinding/html/index.html#memory

File: libraries/AP_ONVIF/AP_ONVIF.cpp [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+    soap_end(soap);
+    return true;
+}
+
+// Generate Random Nonce value
+void AP_ONVIF::rand_nonce(char *nonce, size_t noncelen)
+{
+    size_t i;
+    uint32_t r = (uint32_t)(hal.util->get_hw_rtc()/1000000ULL);
+    (void)memcpy((void *)nonce, (const void *)&r, 4);
```
[the remark this answers is not in the document]
Comment (bugobliterator) [reply libraries/AP_ONVIF/AP_ONVIF.cpp:316]: its suppose to be 4, fixed checking size atleast 4

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c [hunk 3]
```diff
@@ -468,3 +468,39 @@ uint32_t stack_free(void *stack_base)
     return ((uint32_t)p) - (uint32_t)stack_base;
 }
 #endif
+
+#ifdef RNG
+static uint32_t stm32_rand_generate(void)
+{
+    static uint32_t last_value = 0;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:475]: this doesn't look to be thread safe

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c [hunk 4]
```diff
@@ -468,3 +468,39 @@ uint32_t stack_free(void *stack_base)
...
+    static uint32_t new_value = 0;
+    uint32_t error_bits = 0;
+    error_bits = RNG_SR_SEIS | RNG_SR_CEIS;
+    while (new_value == last_value) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:479]: this looks dangerous - under what conditions can this loop forever?

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c [hunk 5]
```diff
@@ -468,3 +468,39 @@ uint32_t stack_free(void *stack_base)
     return ((uint32_t)p) - (uint32_t)stack_base;
 }
 #endif
+
+#ifdef RNG
+static uint32_t stm32_rand_generate(void)
+{
+    static uint32_t last_value = 0;
+    static uint32_t new_value = 0;
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:476]: doesn't need to be static for new_value

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c (hunk 4 above)
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:479]: should be do {} while() not while ()

File: libraries/AP_HAL_Linux/Util.cpp [hunk 6]
```diff
@@ -287,3 +287,27 @@ void *Util::heap_realloc(void *h, void *ptr, size_t new_size)
...
 #endif // ENABLE_HEAP
+
+/**
+ * This method will read random values with set size. Please be aware that this method will block until
+ * the requested amount of data is read at proper entropy level, only use if true randomness is desired
+ * otherwise AP_Math's get_random16() should be used.
+ */
+bool Util::get_random_vals(uint8_t* data, size_t size)
+{
+    int dev_random = open("/dev/random", O_RDONLY);
```
Comment (tridge) [inline libraries/AP_HAL_Linux/Util.cpp:298]: use /dev/urandom on linux

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c [hunk 7]
```diff
@@ -468,3 +468,39 @@ uint32_t stack_free(void *stack_base)
     return ((uint32_t)p) - (uint32_t)stack_base;
 }
 #endif
+
+#ifdef RNG
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:472]: should have some way to enable/disable this feature, perhaps off by default?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c (hunk 3 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:475]: ChibiOS changes moved to separate PR https://github.com/ArduPilot/ardupilot/pull/17855

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c (hunk 4 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:479]: ChibiOS changes moved to separate PR https://github.com/ArduPilot/ardupilot/pull/17855

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c (hunk 5 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:476]: ChibiOS changes moved to separate PR https://github.com/ArduPilot/ardupilot/pull/17855

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c (hunk 4 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:479]: ChibiOS changes moved to separate PR https://github.com/ArduPilot/ardupilot/pull/17855

File: libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c (hunk 7 above)
Comment (bugobliterator) [reply libraries/AP_HAL_ChibiOS/hwdef/common/stm32_util.c:472]: ChibiOS changes moved to separate PR https://github.com/ArduPilot/ardupilot/pull/17855

File: libraries/AP_HAL_Linux/Util.cpp (hunk 6 above)
Comment (bugobliterator) [reply libraries/AP_HAL_Linux/Util.cpp]: Fixed

**Review by tridge — APPROVED** [review]

**Pull request merged**